jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutorWithPool.java

```diff
@@ -0,0 +1,184 @@
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+@ManagedObject("An alternate pool for reserved threads")
+public class ReservedThreadExecutorWithPool extends AbstractLifeCycle implements TryExecutor
+{
+    private static final Logger LOG = Log.getLogger(ReservedThreadExecutorWithPool.class);
+
+    private final Pool<ReservedThread> stack;
+    private final Executor source;
+    private final int capacity;
+
+    private long idleTime = 1L;
+    private TimeUnit idleTimeUnit = TimeUnit.MINUTES;
+
+    public ReservedThreadExecutorWithPool(Executor source, int capacity)
+    {
+        this.stack = new Pool<>(Pool.StrategyType.FIRST, capacity);
+        this.source = source;
+        this.capacity = capacity;
+    }
+
+    @ManagedAttribute(value = "idletimeout in MS", readonly = true)
+    public long getIdleTimeoutMs()
+    {
+        if (idleTimeUnit == null)
+            return 0;
+        return idleTimeUnit.toMillis(idleTime);
+    }
+
+    /**
+     * Set the idle timeout for shrinking the reserved thread pool
+     *
+     * @param idleTime Time to wait before shrinking, or 0 for no timeout.
+     * @param idleTimeUnit Time units for idle timeout
+     */
+    public void setIdleTimeout(long idleTime, TimeUnit idleTimeUnit)
+    {
+        if (isRunning())
+            throw new IllegalStateException();
+        this.idleTime = idleTime;
+        this.idleTimeUnit = idleTimeUnit;
+    }
+
+    @Override
+    public boolean tryExecute(Runnable task)
+    {
+        Pool<ReservedThread>.Entry entry = stack.acquire();
+        ReservedThread reservedThread = entry == null ? null : entry.getPooled();
+        if (reservedThread != null)
+        {
+            return reservedThread.offer(task);
+        }
+        else
+        {
+            if (stack.size() < capacity)
+            {
+                Pool<ReservedThread>.Entry reservedEntry = stack.reserve();
+                if (reservedEntry != null)
+                    source.execute(new ReservedThread(reservedEntry));
+            }
+            return false;
+        }
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        stack.close();
+    }
+
+    private class ReservedThread implements Runnable, Closeable
+    {
+        private final Exchanger<Runnable> exchanger = new Exchanger<>();
+        private final Pool<ReservedThread>.Entry entry;
+
+        public ReservedThread(Pool<ReservedThread>.Entry entry)
+        {
+            this.entry = entry;
+        }
+
+        @Override
+        public void close() throws IOException
+        {
+            try
+            {
+                exchanger.exchange(STOP);
+            }
+            catch (InterruptedException e)
+            {
+                // TODO handle
+            }
+        }
+
+        public boolean offer(Runnable task)
+        {
+            try
+            {
+                exchanger.exchange(task, 1, TimeUnit.SECONDS);
```

Comment:
This is just using a different data structure for the offer, when we don't know if the problem was with the `SynchronousQueue` or not.
You'd have to think that `Exchanger` are internally going to be moderately similar, but `Exchanger` is likely to be more complex because it is unnecessarily bi-directional.

Besides, having the offering thread block (or even yield in an internal spin) here defeats the whole purpose of EPC.   We can't give an opportunity for the calling thread to be moved to another CPU core.